core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -227,55 +166,153 @@ public Schema getOutputSchema(GetSchemaRequest request) {
-    Map<String, Schema> inputs = new TreeMap<>(inputSchemas);
-    Map<String, Schema.Field> outputFields = new LinkedHashMap<>();
-    Map<String, String> duplicateFieldsMap = new HashMap<>();
-
-    for (Map.Entry<String, Schema> input : inputs.entrySet()) {
-      String stageName = input.getKey();
-      Schema inputSchema = input.getValue();
-      Map<String, String> fieldsToSelectRename = perStageFieldsToRename.row(stageName);
-
-      for (Schema.Field inputField : inputSchema.getFields()) {
-        String inputFieldName = inputField.getName();
-        // drop the field if not part of fieldsToSelectRename
-        if (!fieldsToSelectRename.containsKey(inputFieldName)) {
-          continue;
-        }
-        // rename the out field
-        String outputFieldName = fieldsToSelectRename.get(inputFieldName);
 
-        // check if there are any duplicate fields
-        if (outputFields.containsKey(outputFieldName)) {
-          duplicateFieldsMap.put(stageName, com.google.common.base.Joiner.on(",")
-            .join(duplicateFieldsMap.get(stageName), outputFieldName));
-          continue;
-        }
+    // stage name to input schema
+    Map<String, Schema> inputs = new HashMap<>(inputSchemas);
+    // Selected Field name to output field info
+    Map<String, OutputFieldInfo> outputFieldInfo = new LinkedHashMap<>();
+    List<String> duplicateAliases = new ArrayList<>();
+
+    // order of fields in output schema will be same as order of selectedFields
+    Set<Table.Cell<String, String, String>> rows = perStageSelectedFields.cellSet();
+    for (Table.Cell<String, String, String> row : rows) {
+      String stageName = row.getRowKey();
+      String inputFieldName = row.getColumnKey();
+      String alias = row.getValue();
+      Schema inputSchema = inputs.get(stageName);
+
+      // TODO As per CDAP-6402 handle cases where input schema is null at configure time.
+      if (inputSchema == null) {
+        throw new IllegalArgumentException(String.format("Input schema for input stage %s can not be null", stageName));
+      }
 
-        // set nullable fields for non-required inputs
-        if (!Iterables.isEmpty(requiredInputs) && Iterables.contains(requiredInputs, stageName)) {
-          outputFields.put(outputFieldName, Schema.Field.of(outputFieldName, inputField.getSchema()));
-        } else {
-          outputFields.put(outputFieldName, Schema.Field.of(outputFieldName, Schema.nullableOf(inputField.getSchema()
-          )));
+      if (outputFieldInfo.containsKey(alias)) {
+        OutputFieldInfo outInfo = outputFieldInfo.get(alias);
+        if (!duplicateAliases.contains(alias)) {
+          duplicateFields.put(outInfo.getStageName(), outInfo.getInputFieldName());
+          duplicateAliases.add(alias);
         }
+        duplicateFields.put(stageName, inputFieldName);
+        continue;
       }
+
+      Schema.Field inputField = inputSchema.getField(inputFieldName);
```

Comment:
need to handle the case where this is null, which happens if the user specifies a field that is not in the input schema.